applications/RomApplication/custom_strategies/lspg_rom_builder_and_solver.h

```diff
@@ -0,0 +1,587 @@
+        if(!conditions.empty())
+        {
+            block_for_each(conditions, assembly_tls_container, 
+                [&](Condition& r_condition, AssemblyTLS& r_thread_prealloc)
+            {
+                CalculateAssembledResiduals(r_condition, rb, r_thread_prealloc, *pScheme, r_current_process_info);
+            });
+        }
+
+        std::stringstream matrix_market_vector_name;
+        matrix_market_vector_name << "R_" << rModelPart.GetProcessInfo()[TIME] << "_" << rModelPart.GetProcessInfo()[NL_ITERATION_NUMBER] <<  ".res.mm";
+        SparseSpaceType::WriteMatrixMarketVector((matrix_market_vector_name.str()).c_str(), rb);
+
+        KRATOS_INFO_IF("LSPGROMBuilderAndSolver", (this->GetEchoLevel() > 0)) << "Write residuals to train Petrov Galerkin time: " << residual_writing_timer.ElapsedSeconds() << std::endl;
+        KRATOS_CATCH("")
+    }
+
+    ///@}
+    ///@name Protected access
+    ///@{
+
+
+    ///@}
+    ///@name Protected inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Protected life cycle
+    ///@{
+    
+private:
+
+    SizeType mNumberOfRomModes;
+
+    ///@}
+    ///@name Private operations 
+    ///@{
+
+    /**
+     * Computes the local contribution of an element or condition for LSPG
+     */
+    template<typename TEntity>
+    void CalculateLocalContributionLSPG(
+        TEntity& rEntity,
+        LSPGSystemMatrixType& rAglobal,
+        LSPGSystemVectorType& rBglobal,
+        AssemblyTLS& rPreAlloc,
+        TSchemeType& rScheme,
+        const ProcessInfo& rCurrentProcessInfo)
+    {
+        if (rEntity.IsDefined(ACTIVE) && rEntity.IsNot(ACTIVE)) return;
+
+        // Calculate elemental contribution
+        rScheme.CalculateSystemContributions(rEntity, rPreAlloc.lhs, rPreAlloc.romB, rPreAlloc.eq_id, rCurrentProcessInfo);
+        rEntity.GetDofList(rPreAlloc.dofs, rCurrentProcessInfo);
+
+        const SizeType ndofs = rPreAlloc.dofs.size();
+        ResizeIfNeeded(rPreAlloc.phiE, ndofs, this->GetNumberOfROMModes());
+        ResizeIfNeeded(rPreAlloc.romA, ndofs, this->GetNumberOfROMModes());
+
+        const auto &r_geom = rEntity.GetGeometry();
+        RomAuxiliaryUtilities::GetPhiElemental(rPreAlloc.phiE, rPreAlloc.dofs, r_geom, this->mMapPhi);
+
+        noalias(rPreAlloc.romA) = prod(rPreAlloc.lhs, rPreAlloc.phiE);
+
+
+        // Assembly
+        for(SizeType row=0; row < ndofs; ++row)
+        {
+            const SizeType global_row = rPreAlloc.eq_id[row];
+
+            double& r_bi = rBglobal(global_row);
+            AtomicAdd(r_bi, rPreAlloc.romB[row]);
+
+            if(rPreAlloc.dofs[row]->IsFixed()) continue;
+
+            for(SizeType col=0; col < this->GetNumberOfROMModes(); ++col)
+            {
+                // const SizeType global_col = rPreAlloc.eq_id[col];
+                const SizeType global_col = col;
+                double& r_Aij = rAglobal(global_row, global_col);
+                AtomicAdd(r_Aij, rPreAlloc.romA(row, col));
+            }

```

Comment:
```suggestion
            if(rPreAlloc.dofs[row]->IsFree())
            {
                for(SizeType col=0; col < this->GetNumberOfROMModes(); ++col)
                {
                    // const SizeType global_col = rPreAlloc.eq_id[col];
                    const SizeType global_col = col;
                    double& r_Aij = rAglobal(global_row, global_col);
                    AtomicAdd(r_Aij, rPreAlloc.romA(row, col));
                }
            }
```
Though equivalent, I think this is more easy to follow.